Add API for open registration between operators and subclasses (and modes)

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #130261
* __->__ #130064

We add torch.library.Library._register_torch_dispatch_rule. Here, a user
can provide us a specific rule to run for a specific
(torch_dispatch_class, operator) pair. The motivation is that a user
might want to extend a subclass/mode but may not have access to the
source code of the subclass/mode.

I'll make this public in a follow-up PR if we think the approach and API
is good.

Keep in mind that many subclasses will likely deliver their own open
registration solution (DTensor has register_sharding_prop_rule and NJT
has register_jagged_op); _register_torch_dispatch_rule is meant as a
catch-all open registration mechanism for when the subclass hasn't
provided anything more specific.

Test Plan:
- new tests

zou3519 committed:
Add API for open registration between operators and subclasses (and modes)

We add torch.library.Library._register_torch_dispatch_rule. Here, a user
can provide us a specific rule to run for a specific
(torch_dispatch_class, operator) pair. The motivation is that a user
might want to extend a subclass/mode but may not have access to the
source code of the subclass/mode.

I'll make this public in a follow-up PR if we think the approach and API
is good.

Keep in mind that many subclasses will likely deliver their own open
registration solution (DTensor has register_sharding_prop_rule and NJT
has register_jagged_op); _register_torch_dispatch_rule is meant as a
catch-all open registration mechanism for when the subclass hasn't
provided anything more specific.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add API for open registration between operators and subclasses (and modes)"


We add torch.library.Library._register_torch_dispatch_rule. Here, a user
can provide us a specific rule to run for a specific
(torch_dispatch_class, operator) pair. The motivation is that a user
might want to extend a subclass/mode but may not have access to the
source code of the subclass/mode.

I'll make this public in a follow-up PR if we think the approach and API
is good.

Keep in mind that many subclasses will likely deliver their own open
registration solution (DTensor has register_sharding_prop_rule and NJT
has register_jagged_op); _register_torch_dispatch_rule is meant as a
catch-all open registration mechanism for when the subclass hasn't
provided anything more specific.

Test Plan:
- new tests

[ghstack-poisoned]

albanD commented on torch/_library/simple_registry.py:
```diff
@@ -42,8 +45,39 @@ class SimpleOperatorEntry:
+    ) -> RegistrationHandle:
+        if self.find(torch_dispatch_class):
+            raise RuntimeError(
+                f"{torch_dispatch_class} already has a `__torch_dispatch__` rule registered"
```
nit: do we have the qualname of the op here we could add to the error message?

albanD commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -298,6 +310,29 @@ static py::object dispatch_on_subclass(
+    if (!is_torch_function) {
+      auto maybe_torch_dispatch_rule = maybe_get_registered_torch_dispatch_rule(
+          torch_api_function, py::reinterpret_borrow<py::object>(arg));
+      if (maybe_torch_dispatch_rule.ptr() != Py_None) {
```
if this is a py::object, you have `.is_none()` IIRC

albanD reviewed (approved):
Sounds great!

zou3519 committed:
Update on "Add API for open registration between operators and subclasses (and modes)"


We add torch.library.Library._register_torch_dispatch_rule. Here, a user
can provide us a specific rule to run for a specific
(torch_dispatch_class, operator) pair. The motivation is that a user
might want to extend a subclass/mode but may not have access to the
source code of the subclass/mode.

I'll make this public in a follow-up PR if we think the approach and API
is good.

Keep in mind that many subclasses will likely deliver their own open
registration solution (DTensor has register_sharding_prop_rule and NJT
has register_jagged_op); _register_torch_dispatch_rule is meant as a
catch-all open registration mechanism for when the subclass hasn't
provided anything more specific.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change but test_profiler_tree is failing in trunk after this lands https://hud.pytorch.org/pytorch/pytorch/commit/922d2737d5e0ad22ee1dcf91c48ab09d641de840, maybe a landrace' -c landrace

zou3519 committed:
Update on "Add API for open registration between operators and subclasses (and modes)"


We add torch.library.Library._register_torch_dispatch_rule. Here, a user
can provide us a specific rule to run for a specific
(torch_dispatch_class, operator) pair. The motivation is that a user
might want to extend a subclass/mode but may not have access to the
source code of the subclass/mode.

I'll make this public in a follow-up PR if we think the approach and API
is good.

Keep in mind that many subclasses will likely deliver their own open
registration solution (DTensor has register_sharding_prop_rule and NJT
has register_jagged_op); _register_torch_dispatch_rule is meant as a
catch-all open registration mechanism for when the subclass hasn't
provided anything more specific.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 commented:
@huydhn afaict that test was added in 2022, is something else wrong with CI? I did confirm that the test failed locally.

XuehaiPan commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -259,6 +259,18 @@ static PyObject* get_type_of_overloaded_arg(PyObject* obj_or_type) {
+static py::object maybe_get_registered_torch_dispatch_rule(
+    PyObject* torch_api_function,
+    const py::object& torch_dispatch_object) {
+  static auto find_torch_dispatch_rule =
+      py::module_::import("torch._library.simple_registry")
+          .attr("find_torch_dispatch_rule");
```
static `py::object` may lead to a segmentation fault.

```suggestion
  PYBIND11_CONSTINIT static py::gil_safe_call_once_and_store<py::object>
      storage;
  py::object find_torch_dispatch_rule =
      storage.call_once_and_store_result([]() -> py::object {
        return py::module_::import("torch._library.simple_registry")
            .attr("find_torch_dispatch_rule");
      });
```

I opened a PR to resolve similar issues in the PyTorch C++ codebase.

- #130341

zou3519 replied:
Thanks, I'll try that

zou3519 committed:
Update on "Add API for open registration between operators and subclasses (and modes)"


We add torch.library.Library._register_torch_dispatch_rule. Here, a user
can provide us a specific rule to run for a specific
(torch_dispatch_class, operator) pair. The motivation is that a user
might want to extend a subclass/mode but may not have access to the
source code of the subclass/mode.

I'll make this public in a follow-up PR if we think the approach and API
is good.

Keep in mind that many subclasses will likely deliver their own open
registration solution (DTensor has register_sharding_prop_rule and NJT
has register_jagged_op); _register_torch_dispatch_rule is meant as a
catch-all open registration mechanism for when the subclass hasn't
provided anything more specific.

Test Plan:
- new tests

[ghstack-poisoned]

huydhn commented:
> @huydhn afaict that test was added in 2022, is something else wrong with CI? I did confirm that the test failed locally.

Oh, I see @clee2000 has added `ci-no-td` to the PR, so I think it's a target determination miss where the test should run but it didn't.  Adding the label will bypass TD and run all tests as usual.

izaitsevfb commented:
@pytorchbot revert -m "fails internal builds, see [D59553526](https://www.internalfb.com/diff/D59553526)" -c ghfirst

```
ERROR    ...
ERROR    reason: Action failed: fbcode//caffe2:_C_impl_cuda (fbcode//buck2/platform/execution:linux-x86_64#22f53947965bda2d) (cxx_compile torch/csrc/utils/python_arg_parser.cpp (pic))
ERROR    Remote command returned non-zero exit code 1
ERROR    Remote action, reproduce with: `frecli cas download-action 9c60783056764f8cb03d98ef26e521ad27a68f14daa5ed7ef3c9d9790a5e70cb:145`
ERROR    Stdout: <empty>
ERROR    Stderr:
ERROR    fbcode/caffe2/torch/csrc/utils/python_arg_parser.cpp:265:3: error: unknown type name 'PYBIND11_CONSTINIT'
ERROR      PYBIND11_CONSTINIT static py::gil_safe_call_once_and_store<py::object>
ERROR      ^
ERROR    fbcode/caffe2/torch/csrc/utils/python_arg_parser.cpp:265:33: error: definition or redeclaration of 'gil_safe_call_once_and_store' not allowed inside a function
ERROR      PYBIND11_CONSTINIT static py::gil_safe_call_once_and_store<py::object>
ERROR                                ~~~~^
ERROR    fbcode/caffe2/torch/csrc/utils/python_arg_parser.cpp:265:33: error: no member named 'gil_safe_call_once_and_store' in namespace 'pybind11'
ERROR      PYBIND11_CONSTINIT static py::gil_safe_call_once_and_store<py::object>
ERROR                                ~~~~^
ERROR    fbcode/caffe2/torch/csrc/utils/python_arg_parser.cpp:265:61: error: expected ';' at end of declaration
ERROR      PYBIND11_CONSTINIT static py::gil_safe_call_once_and_store<py::object>
ERROR                                                                ^
ERROR                                                                ;
ERROR    fbcode/caffe2/torch/csrc/utils/python_arg_parser.cpp:268:7: error: use of undeclared identifier 'storage'
ERROR          storage
ERROR          ^
ERROR    5 errors generated.
ERROR 
ERROR      --> fbcode/buck2/prelude/python/sourcedb/build.bxl:25:9
ERROR       |
ERROR    25 |         fail(error_message)
ERROR       |         ^^^^^^^^^^^^^^^^^^^
ERROR       |
ERROR 
```

zou3519 committed:
Update on "Add API for open registration between operators and subclasses (and modes)"


We add torch.library.Library._register_torch_dispatch_rule. Here, a user
can provide us a specific rule to run for a specific
(torch_dispatch_class, operator) pair. The motivation is that a user
might want to extend a subclass/mode but may not have access to the
source code of the subclass/mode.

I'll make this public in a follow-up PR if we think the approach and API
is good.

Keep in mind that many subclasses will likely deliver their own open
registration solution (DTensor has register_sharding_prop_rule and NJT
has register_jagged_op); _register_torch_dispatch_rule is meant as a
catch-all open registration mechanism for when the subclass hasn't
provided anything more specific.

Test Plan:
- new tests

[ghstack-poisoned]

XuehaiPan commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -259,6 +259,18 @@ static PyObject* get_type_of_overloaded_arg(PyObject* obj_or_type) {
+static py::object maybe_get_registered_torch_dispatch_rule(
+    PyObject* torch_api_function,
+    const py::object& torch_dispatch_object) {
+  static auto find_torch_dispatch_rule =
+      py::module_::import("torch._library.simple_registry")
+          .attr("find_torch_dispatch_rule");
```
```suggestion
  static const py::handle find_torch_dispatch_rule =
      py::module_::import("torch._library.simple_registry")
          .attr("find_torch_dispatch_rule").release();
```

zou3519 replied:
What's the difference?
Also I can fix this up (go back to the original approach you suggested) when we update pybind11 internally (which will hopefully be soon); I want to merge this PR sooner than later because it may have performance implications.

XuehaiPan replied:
`py::handle` is equivalent to `PyObject *`. It will not do `Py_DECREF(obj.m_ptr)` on C++ instance destruction where `py::object` will. `handle = object.release()` will leak the instance to keep it alive on Python side.

zou3519 committed:
Update on "Add API for open registration between operators and subclasses (and modes)"


We add torch.library.Library._register_torch_dispatch_rule. Here, a user
can provide us a specific rule to run for a specific
(torch_dispatch_class, operator) pair. The motivation is that a user
might want to extend a subclass/mode but may not have access to the
source code of the subclass/mode.

I'll make this public in a follow-up PR if we think the approach and API
is good.

Keep in mind that many subclasses will likely deliver their own open
registration solution (DTensor has register_sharding_prop_rule and NJT
has register_jagged_op); _register_torch_dispatch_rule is meant as a
catch-all open registration mechanism for when the subclass hasn't
provided anything more specific.

Test Plan:
- new tests

[ghstack-poisoned]
